controllers/metal3datatemplate_controller.go

```
@@ -0,0 +1,176 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	dataTemplateControllerName = "Metal3DataTemplate-controller"
+)
+
+// Metal3DataTemplateReconciler reconciles a Metal3DataTemplate object
+type Metal3DataTemplateReconciler struct {
+	Client         client.Client
+	ManagerFactory baremetal.ManagerFactoryInterface
+	Log            logr.Logger
+}
+
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=metal3datatemplates,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=metal3datatemplates/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=events,verbs=get;list;watch;create;update;patch
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile handles Metal3Machine events
+func (r *Metal3DataTemplateReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, rerr error) {
+	ctx := context.Background()
+	metadataLog := r.Log.WithName(dataTemplateControllerName).WithValues("metal3-datatemplate", req.NamespacedName)
+
+	// Fetch the Metal3DataTemplate instance.
+	capm3Metadata := &capm3.Metal3DataTemplate{}
+
+	if err := r.Client.Get(ctx, req.NamespacedName, capm3Metadata); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	helper, err := patch.NewHelper(capm3Metadata, r.Client)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to init patch helper")
+	}
+	// Always patch capm3Machine exiting this function so we can persist any Metal3Machine changes.
+	defer func() {
+		err := helper.Patch(ctx, capm3Metadata)
+		if err != nil {
+			metadataLog.Info("failed to Patch capm3Metadata")
+		}
+	}()
+
+	// Fetch the Cluster.
+	cluster, err := util.GetClusterFromMetadata(ctx, r.Client, capm3Metadata.ObjectMeta)
+	if capm3Metadata.ObjectMeta.DeletionTimestamp.IsZero() {
```

this condition is also checked at line 107 for reconciling delete. I am wondering the need to check this multiple times. Also in case there is deletion timestamp set, will it then ignore the err if for example when the cluster label was missing?

---

yes, the idea is that in a case of deletion, we ignore a failure on getting the cluster. Otherwise we could be blocking here if the cluster was removed already.